### PR TITLE
Search Roadmap and Architecture

### PR DESCRIPTION
**Pull request by emmaindal**
## Hey, I just made a Pull Request!

Roadmap and documentation for Backstage Global Search. Also includes the search architecture proposal. 

- [x]  Closes: #4008 

- [x] Closes: #4009 

Screen Shot 2021-01-14 at 3 20 03 PM

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)

### CODE REVIEW

File: docs/features/search/README.md
Comment (OrkoHunter) [inline docs/features/search/README.md:2]: Should also update `mkdocs.yml` and `microsite/sidebars.json` for it to show up on the docs site?

File: docs/features/search/README.md
Comment (OrkoHunter) [inline docs/features/search/README.md:10]: Should we call it `Backstage Global Search` or just `Backstage Search`. Thinking in terms of a new user discovering Backstage for the first time.

File: docs/features/search/README.md
Comment (OrkoHunter) [inline docs/features/search/README.md:47]: These can just be `User should be able to...`, but maybe not.

File: docs/features/search/README.md
Comment (OrkoHunter) [inline docs/features/search/README.md:61]: I think we should use either `app config` or `app_config.yaml`. :)

File: docs/features/search/README.md
Comment (emmaindal) [reply docs/features/search/README.md:10]: good question, i've mixed them both 😆, but yes we should go for one. Let's go for Backstage Search to avoid confusion.

File: docs/features/search/README.md
Comment (OrkoHunter) [inline docs/features/search/README.md:81]: Just a nit - Can we use the phrase `Architecture diagram` over `Big picture` ? IMO one is more widely used phrase.

File: docs/features/search/README.md
Comment (adamdmharvey) [reply docs/features/search/README.md:47]: This is leaning towards the traditional story outline of:

* As a `<user/persona>`...
* I should/would like to `<do this thing>`...
* So that I can `<obtain this result/value>`

I sort of like it as is, with the potential to add more reasoning for why a user would want to do any of those things at a later date? (the why helping the dev, esp open source ones, understand why they are doing what they are doing, other than someone else just gave them a TODO to blow through)

File: docs/features/search/README.md
Comment (OrkoHunter) [inline docs/features/search/README.md:85]: Nice! Looking forward to what more the community asks of us. `Solr` comes to mind.

File: docs/features/search/README.md
Comment (OrkoHunter) [inline docs/features/search/README.md:87]:
```suggestion
[Reach out to us](#feedback) if you want to chat about support for more search
```
?

Comment (emmaindal) [reply docs/features/search/README.md:87]: yesss thanks, that was just me deciding whether or not having the feedback section here

File: docs/features/search/README.md
Comment (emmaindal) [reply docs/features/search/README.md:85]: +1

File: docs/features/search/README.md
Comment (adamdmharvey) [reply docs/features/search/README.md:47]: Also to note, there may be some of those which may say "As a Backstage integrator, I should be able to..." as there may be features that relate to how to keep the search index healthy or understand how it's working, as opposed to just the end user as well?

Comment (emmaindal) [reply docs/features/search/README.md:47]: would say +1 to the above and yes, there is more potential to add more around the third bullet point.

Comment (OrkoHunter) [reply docs/features/search/README.md:47]: As a user
As a plugin developer
As a Backstage adopter
...
So many stories! :)

Makes sense to keep it. 👍

Comment (emmaindal) [reply docs/features/search/README.md:47]: @adamdmharvey Yeah we have some already saying "As a app developer" or "As a plugin developer", would you prefer "As a Backstage integrator" instead?

Comment (adamdmharvey) [reply docs/features/search/README.md:47]: @emmaindal I think so, only in so much as I've seen that being used a lot and I think it makes sense to me. I think to align it well with https://backstage.io/docs/overview/adopting, although even that page doesn't call it Integrator.

I do see this in the FAQ that lists three roles:
https://backstage.io/docs/FAQ#what-is-the-end-to-end-user-flow-the-happy-path-story

So maybe trying to keep all those consistent. (even if it means updating the FAQ, as "contributor" is also similar to your example of "plugin developer", etc.)

Comment (emmaindal) [reply docs/features/search/README.md:47]: @adamdmharvey Great thanks for linking that question from the FAQ and express your thinking!

**Review by iamEAP — APPROVED** [review]

**Comment by emmaindal** [comment]
@backstage/maintainers hey, we need your approval as we are adding some new docs! 👀

File: docs/features/search/architecture.md
Comment (soapraj) [inline docs/features/search/architecture.md:32]: Hey @emmaindal - does this mean a plugin can expose any index any kind of data? When you say content is it new entities or metadata about existing entities(covered by the next point) or any data?

Comment (emmaindal) [reply docs/features/search/architecture.md:32]: Hey @soapraj, thanks for feedback! our thinking is any data, as TechDocs for example will index documentation content for example which is outside of entity/metadata. Maybe its good to clarify though!

File: docs/features/search/README.md
Comment (freben) [inline docs/features/search/README.md:5]: insert `# prettier-ignore` between line 4 and 5, and make the description a oneliner - otherwise it will get cut off

**Review by OrkoHunter — APPROVED** [review]

**Comment by emmaindal** [comment]
@backstage/maintainers would be nice if you can approve this!

**Comment by Rugvip** [comment]
@iamEAP Possibly, might also be really annoying to get working if you have authentication or any other kind of non-standard setup in the app.

**Pull request merged**
